### PR TITLE
Enable cuav neo v2 pro safety state

### PR DESCRIPTION
**Pull request by gitfishup**
Enable cuav neo v2 pro safety state.

### CODE REVIEW

**Comment by dagar** [comment]
Can you clarify the messages? Where does `hex/equipment/indication/20000.SafetyStateCommand.uavcan` come from?

**Comment by gitfishup** [comment]
> Can you clarify the messages? Where does `hex/equipment/indication/20000.SafetyStateCommand.uavcan` come from?

Base on ardupilot :https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_UAVCAN/dsdl/ardupilot/indication/20000.SafetyState.uavcan

**Comment by dagar** [comment]
> > Can you clarify the messages? Where does `hex/equipment/indication/20000.SafetyStateCommand.uavcan` come from?
> 
> Base on ardupilot :https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_UAVCAN/dsdl/ardupilot/indication/20000.SafetyState.uavcan

The message should be copied identically (enums are missing) with the namespace maintained (this isn't Hex DSDL).

Before we can do that we need to get clarification of the licensing. I think we have to assume these files are GPLv3 like the rest of the project, so we can't just copy them into PX4 (BSD 3 license).

**Comment by gitfishup** [comment]
> > > Can you clarify the messages? Where does `hex/equipment/indication/20000.SafetyStateCommand.uavcan` come from?
> > 
> > 
> > Base on ardupilot :https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_UAVCAN/dsdl/ardupilot/indication/20000.SafetyState.uavcan
> 
> The message should be copied identically (enums are missing) with the namespace maintained (this isn't Hex DSDL).
> 
> Before we can do that we need to get clarification of the licensing. I think we have to assume these files are GPLv3 like the rest of the project, so we can't just copy them into PX4 (BSD 3 license).
> 
> #13893
Can we do this if the ArduPilot manager contributes to PX4?

**Comment by tridge** [comment]
> Before we can do that we need to get clarification of the licensing. I think we have to assume these files are GPLv3 like the rest of the project, so we can't just copy them into PX4 (BSD 3 license).

I am the author of that specific dsdl file. You may use it in PX4 under BSD 3 license.

**Comment by dagar** [comment]
Thanks @tridge.

@gitfishup this should be good to merge if you pull in a copy of 20000.SafetyState.uavcan, update the naming/path, and use the STATUS_SAFETY enums instead of magic numbers.

**Comment by gitfishup** [comment]
> Thanks @tridge.
> 
> @gitfishup this should be good to merge if you pull in a copy of 20000.SafetyState.uavcan, update the naming/path, and use the STATUS_SAFETY enums instead of magic numbers.

How about this pull request?

**Comment by gitfishup** [comment]
> Thanks @tridge.
> 
> @gitfishup this should be good to merge if you pull in a copy of 20000.SafetyState.uavcan, update the naming/path, and use the STATUS_SAFETY enums instead of magic numbers.

How about this pull request?@dagar

**Comment by dagar** [comment]
> > Thanks @tridge.
> > @gitfishup this should be good to merge if you pull in a copy of 20000.SafetyState.uavcan, update the naming/path, and use the STATUS_SAFETY enums instead of magic numbers.
> 
> How about this pull request?@dagar

It shouldn't be in the Hex namespace.

Move `src/drivers/uavcan/dsdl/com/hex/equipment/indication/20000.SafetyState.uavcan` -> `src/drivers/uavcan/dsdl/ardupilot/indication/20000.SafetyState.uavcan` 

Then you'll need to update things like `com::hex::equipment::indication::SafetyState cmd` -> `ardupilot::indication::SafetyState cmd`.

**Comment by gitfishup** [comment]
> > > Thanks @tridge.
> > > @gitfishup this should be good to merge if you pull in a copy of 20000.SafetyState.uavcan, update the naming/path, and use the STATUS_SAFETY enums instead of magic numbers.
> > 
> > 
> > How about this pull request?@dagar
> 
> It shouldn't be in the Hex namespace.
> 
> Move `src/drivers/uavcan/dsdl/com/hex/equipment/indication/20000.SafetyState.uavcan` -> `src/drivers/uavcan/dsdl/ardupilot/indication/20000.SafetyState.uavcan`
> 
> Then you'll need to update things like `com::hex::equipment::indication::SafetyState cmd` -> `ardupilot::indication::SafetyState cmd`.

Thanks.I have changed that and pull request again.

File: src/drivers/uavcan/safety_state.cpp [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+	return 0;
+}
+
+void UavcanSafetyState::periodic_update(const uavcan::TimerEvent &)
+{
+	if (_actuator_armed_sub.updated()) {
+		_actuator_armed_sub.copy(&_actuator_armed);
+		ardupilot::indication::SafetyState cmd;
+
```
Comment (dagar) [inline src/drivers/uavcan/safety_state.cpp:66]: You don't need to store `_actuator_armed`.

```suggestion
	actuator_armed_s actuator_armed;
	if (_actuator_armed_sub.update(&actuator_armed)) {
		ardupilot::indication::SafetyState cmd;

```

File: src/drivers/uavcan/safety_state.hpp [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+	/*
+	 * Publish CAN Safety state led
+	 */
+	uavcan::Publisher<ardupilot::indication::SafetyState> _safety_state_pub;
+
+	uavcan::TimerEventForwarder<TimerCbBinder> _timer;
+
+	uORB::Subscription _actuator_armed_sub{ORB_ID(actuator_armed)};
+	actuator_armed_s _actuator_armed{};
```
Comment (dagar) [inline src/drivers/uavcan/safety_state.hpp:82]:
```suggestion
```

File: src/drivers/uavcan/safety_state.cpp [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+		if (_actuator_armed.prearmed == true) {
```
Comment (dagar) [inline src/drivers/uavcan/safety_state.cpp:67]: Is prearmed actually what you want here? Prearmed is an optional state before armed where things like servos can be active, but not ESCs.


Would this make sense?

```suggestion
		if (actuator_armed.armed || actuator_armed.prearmed) {
```

Comment (gitfishup) [reply src/drivers/uavcan/safety_state.cpp]: > Is prearmed actually what you want here? Prearmed is an optional state before armed where things like servos can be active, but not ESCs.
> 
> Would this make sense?

This make sense. Thanks a lot .

**Review by dagar — APPROVED** [review]

**Pull request merged**